sql: add dump subcommand

madelynnblue opened the pull request:
Introspect over a specified table to determine its structure. Iteratively
select all rows from the table in pages using time travel queries to
guarantee they don't change.

TestDumpRandom was run for 20 minutes and didn't report any unexpected
values, so I have fairly high confidence that the bytes a user has stored
will correctly be inserted if the dump file is piped into the sql subcommand.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/7511)

<!-- Reviewable:end -->

tamird commented:
Reviewed 2 of 5 files at r1, 3 of 5 files at r2.
Review status: 5 of 7 files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

*[cli/dump.go, line 55 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7511#-KLW-hBVZ_u5yxjHI0Eq:-KLW-hBVZ_u5yxjHI0Er:-947416919) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d4eb7c7183ff1476bae4df7104fe683be2fe3ec/cli/dump.go#L55)):*
> ```Go
> 	defer conn.Close()
> 
> 	w := os.Stdout
> ```

random local variable

---

*[cli/dump.go, line 56 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7511#-KLW-jARdFMn88xb2ZnD:-KLW-jARdFMn88xb2ZnE:869081535) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d4eb7c7183ff1476bae4df7104fe683be2fe3ec/cli/dump.go#L56)):*
> ```Go
> 
> 	w := os.Stdout
> 	if err := dumpTable(w, conn, args[0], args[1]); err != nil {
> ```

return dumpTable...

---

*[cli/dump.go, line 76 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7511#-KLVzWgF_DdQceBLB20K:-KLVzWgF_DdQceBLB20L:-1890139656) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d4eb7c7183ff1476bae4df7104fe683be2fe3ec/cli/dump.go#L76)):*
> ```Go
> 	// Fetch all table metadata in a transaction and transaction time to guarantee
> 	// it doesn't change between the various SHOW statements.
> 	if err := conn.Exec("BEGIN", nil); err != nil {
> ```

why doesn't this use conn.Begin?

---

*[cli/dump.go, line 90 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7511#-KLW02tJ7wbSpNpE7Tvg:-KLW02tJ7wbSpNpE7Tvh:1598690509) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d4eb7c7183ff1476bae4df7104fe683be2fe3ec/cli/dump.go#L90)):*
> ```Go
> 
> 	// Fetch column types.
> 	rows, err := conn.Query(fmt.Sprintf("SHOW COLUMNS FROM %s", tablename), nil)
> ```

wouldn't it be easier to query system tables and unmarshal protos?

---

*[cli/dump.go, line 104 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7511#-KLW-wjoGanEVy8D4sjK:-KLW-wjoGanEVy8D4sjL:776107439) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d4eb7c7183ff1476bae4df7104fe683be2fe3ec/cli/dump.go#L104)):*
> ```Go
> 		b, ok := vals[0].([]byte)
> 		if !ok {
> 			return fmt.Errorf("unexpected value: %T", vals[1])
> ```

vals[1]?

---

*[cli/dump.go, line 109 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7511#-KLW-xuL0f6qZCHEmX0Z:-KLW-xuL0f6qZCHEmX0_:776110322) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d4eb7c7183ff1476bae4df7104fe683be2fe3ec/cli/dump.go#L109)):*
> ```Go
> 		b, ok = vals[1].([]byte)
> 		if !ok {
> 			return fmt.Errorf("unexpected value: %T", vals[4])
> ```

vals[4]?

---

*[cli/sql_util.go, line 81 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7511#-KLW-M3dE6efHYnW0qIZ:-KLW-M3dE6efHYnW0qI_:1135948951) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d4eb7c7183ff1476bae4df7104fe683be2fe3ec/cli/sql_util.go#L81)):*
> ```Go
> }
> 
> func (c *sqlConn) QueryRow(query string, args []driver.Value) ([]driver.Value, error) {
> ```

I think this wants to be `args ...driver.Value`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7511)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 5 of 7 files reviewed at latest revision, 12 unresolved discussions, all commit checks successful.

---

*[cli/dump.go, line 34 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7511#-KLNjd2AY3lw7bySVS0o:-KLNjd2BLrCShzHXgLrT:399626490) ([raw file](https://github.com/cockroachdb/cockroach/blob/8cd9b7b9426d4c369c7da88f54fc15880c3c96fe/cli/dump.go#L34)):*
> ```Go
> // dumpCmd dumps SQL tables.
> var dumpCmd = &cobra.Command{
> 	Use:   "dump [options] <database> <table>",
> ```

Should this be `<database>.<table>` to match up with qualified table specification in SQL? 

---

*[cli/dump.go, line 55 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7511#-KLNjMpG7bloRRr8tNWY:-KLNjMpG7bloRRr8tNWZ:920625545) ([raw file](https://github.com/cockroachdb/cockroach/blob/8cd9b7b9426d4c369c7da88f54fc15880c3c96fe/cli/dump.go#L55)):*
> ```Go
> 	defer conn.Close()
> 
> 	w := os.Stdout
> ```

The `w` variable isn't providing much value. Why not just pass `os.Stdout` as the first parameter to `dumpTable`?

---

*[cli/dump.go, line 89 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7511#-KLNlmVR7HsXFnS3jdRT:-KLNlmVR7HsXFnS3jdRU:-806570312) ([raw file](https://github.com/cockroachdb/cockroach/blob/8cd9b7b9426d4c369c7da88f54fc15880c3c96fe/cli/dump.go#L89)):*
> ```Go
> 	clusterTS := time.Unix(0, clusterTSStart).Format(time.RFC3339Nano)
> 
> 	// Fetch column types.
> ```

Given this and the code to fetch the primary key below, I wonder if it would be better to read the table descriptor from the `system.descriptors` table. The query would be mildly complicated, but not horrendously so. We have code in `zone.go` to do queries against those tables already, so you'd be following existing patterns.

---

*[cli/dump.go, line 154 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7511#-KLNkAeqDPbC5mnLpUbO:-KLNkAerLtckVQBr4LYR:-1728411681) ([raw file](https://github.com/cockroachdb/cockroach/blob/8cd9b7b9426d4c369c7da88f54fc15880c3c96fe/cli/dump.go#L154)):*
> ```Go
> 
> 	// Build the SELECT query.
> 	sbuf := bytes.NewBufferString(fmt.Sprintf("SELECT %s, * FROM %s AS OF SYSTEM TIME '%s'", indexes, tablename, clusterTS))
> ```

This is selecting the primary key columns and all of the columns, so the primary key columns are duplicated, right?

Should we be using the `<table>@<index>` syntax to force using the primary key index?

---

*[cli/dump_test.go, line 168 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7511#-KLNnLpcbkO9hZGbGENh:-KLNnLpcbkO9hZGbGENi:96160545) ([raw file](https://github.com/cockroachdb/cockroach/blob/8cd9b7b9426d4c369c7da88f54fc15880c3c96fe/cli/dump_test.go#L168)):*
> ```Go
> // This function is disabled because it is not a deterministic test and
> // runs forever.
> func _TestDumpRandom(t *testing.T) {
> ```

Could you make it run for a period of time that is controlled by a flag with a default of 1s?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7511)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Reviewed 2 of 5 files at r1, 5 of 5 files at r2.
Review status: all files reviewed at latest revision, 16 unresolved discussions, all commit checks successful.

---

*[cli/dump.go, line 74 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7511#-KLXVm0CkD-Gs_tUiKYU:-KLXVm0CkD-Gs_tUiKYV:-386014512) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d4eb7c7183ff1476bae4df7104fe683be2fe3ec/cli/dump.go#L74)):*
> ```Go
> 	}
> 
> 	// Fetch all table metadata in a transaction and transaction time to guarantee
> ```

missing verb.

---

*[cli/dump.go, line 206 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7511#-KLXWHO52f9u9izWOes_:-KLXWHO6sbXKaOX1DhSU:2008323615) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d4eb7c7183ff1476bae4df7104fe683be2fe3ec/cli/dump.go#L206)):*
> ```Go
> 
> 	if clusterTSStart != clusterTSEnd {
> 		return fmt.Errorf("unexpected transaction timestamp change")
> ```

Good assertion, but this is really not possible unless there's a bug.

---

*[cli/dump.go, line 247 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7511#-KLXXS7E_5mFt4PvJ5jC:-KLXXS7E_5mFt4PvJ5jD:-147853596) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d4eb7c7183ff1476bae4df7104fe683be2fe3ec/cli/dump.go#L247)):*
> ```Go
> 			ivals := make([]string, len(vals))
> 			// Values need to be correctly encoded for INSERT statements in a text file.
> 			for si, sv := range vals {
> ```

The part of the code below makes me wonder why there isn't a standardized facility through which values (given a coltype) represent themselves as a string which can be used in SQL inserts.

---

*[cli/dump_test.go, line 168 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7511#-KLNnLpcbkO9hZGbGENh:-KLXWmVb4y1STz_7OY_4:819706989) ([raw file](https://github.com/cockroachdb/cockroach/blob/8cd9b7b9426d4c369c7da88f54fc15880c3c96fe/cli/dump_test.go#L168)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Could you make it run for a period of time that is controlled by a flag with a default of 1s?
</details>
Looks like the comment is simply stale. It runs 10 iterations, doesn't it?

---

*[cli/dump_test.go, line 189 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7511#-KLXWrp97QtT7omAmQLd:-KLXWrp97QtT7omAmQLe:-1249278824) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d4eb7c7183ff1476bae4df7104fe683be2fe3ec/cli/dump_test.go#L189)):*
> ```Go
> 	}
> 
> 	// Generate a random seed. It is saved and printed in case of an error.
> ```

`rng, seed := randutil.NewPseudoRand()`

I don't know here you print it. I'd just print it right here via `log.Infof` (not `t.Logf`, these can get lost if a panic happens iirc).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7511)*
<!-- Sent from Reviewable.io -->

petermattis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> modulo open comments.

---

Review status: 5 of 7 files reviewed at latest revision, 17 unresolved discussions, some commit checks failed.

---

*[cli/dump.go, line 76 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7511#-KLVzWgF_DdQceBLB20K:-KLYrf86asWPvpqXCeE9:-119661878) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d4eb7c7183ff1476bae4df7104fe683be2fe3ec/cli/dump.go#L76)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> why doesn't this use conn.Begin?
</details>
This is an `*sqlConn`, not a `database/sql.DB`.

---

*[cli/dump.go, line 93 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7511#-KLYrs0Y9serMSO1PHoq:-KLYrs0Y9serMSO1PHor:-481806823) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ccd39cab449d9dc326eeb9c56bbe64d8918b681/cli/dump.go#L93)):*
> ```Go
> 		JOIN system.namespace dbs
> 			ON dbs.id = tables.parentid
> 		WHERE tables.name=$1 AND dbs.name=$2`, []driver.Value{origTableName, origDBName})
> ```

Whoa!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7511#-:-KLYsmckNRiisj67qBA6:831880511)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: 5 of 7 files reviewed at latest revision, 17 unresolved discussions, some commit checks failed.

---

*[cli/dump.go, line 76 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7511#-KLVzWgF_DdQceBLB20K:-KLZ446yrVGFiAzw6SWo:585958656) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d4eb7c7183ff1476bae4df7104fe683be2fe3ec/cli/dump.go#L76)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> This is an `*sqlConn`, not a `database/sql.DB`.
</details>
sqlConn.conn is a sqlConnI which implements drver.Conn, which has a `Begin` method. What'd I miss?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7511)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 5 of 7 files reviewed at latest revision, 17 unresolved discussions, some commit checks failed.

---

*[cli/dump.go, line 76 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7511#-KLVzWgF_DdQceBLB20K:-KLZ5pnKqGPHLtHS2oIM:-1841813380) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d4eb7c7183ff1476bae4df7104fe683be2fe3ec/cli/dump.go#L76)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> sqlConn.conn is a sqlConnI which implements drver.Conn, which has a `Begin` method. What'd I miss?
</details>
You're correct that there is an `sqlConn.conn.Begin` method, but `conn` is an `*sqlConn` and there is no `sqlConn.Begin` method. We could add one, but it wouldn't be providing much, if anything, over sending an explicit `BEGIN` statement.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7511)*
<!-- Sent from Reviewable.io -->

mjibson commented:
Review status: 5 of 7 files reviewed at latest revision, 17 unresolved discussions, some commit checks pending.

---

*[cli/dump.go, line 34 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7511#-KLNjd2AY3lw7bySVS0o:-KLXQ6G86QBoDlvlYZ1d:-209922296) ([raw file](https://github.com/cockroachdb/cockroach/blob/8cd9b7b9426d4c369c7da88f54fc15880c3c96fe/cli/dump.go#L34)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Should this be `<database>.<table>` to match up with qualified table specification in SQL? 
</details>
In the future I think omitting the table name will dump an entire database, so leaving it as a separate argument makes it easy to detect that case. pg_dump also separates out database and table names, so I think this is fine.

---

*[cli/dump.go, line 55 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7511#-KLNjMpG7bloRRr8tNWY:-KLXQgxk8ZX4HHz6JboL:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/8cd9b7b9426d4c369c7da88f54fc15880c3c96fe/cli/dump.go#L55)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> The `w` variable isn't providing much value. Why not just pass `os.Stdout` as the first parameter to `dumpTable`?
</details>
Done.

---

*[cli/dump.go, line 89 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7511#-KLNlmVR7HsXFnS3jdRT:-KLa9D_QeTA-wSxrBgTF:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/8cd9b7b9426d4c369c7da88f54fc15880c3c96fe/cli/dump.go#L89)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Given this and the code to fetch the primary key below, I wonder if it would be better to read the table descriptor from the `system.descriptors` table. The query would be mildly complicated, but not horrendously so. We have code in `zone.go` to do queries against those tables already, so you'd be following existing patterns.
</details>
Done.

---

*[cli/dump.go, line 154 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7511#-KLNkAeqDPbC5mnLpUbO:-KLXdAgRcHNeQn50cEBW:1775281126) ([raw file](https://github.com/cockroachdb/cockroach/blob/8cd9b7b9426d4c369c7da88f54fc15880c3c96fe/cli/dump.go#L154)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> This is selecting the primary key columns and all of the columns, so the primary key columns are duplicated, right?
> 
> Should we be using the `<table>@<index>` syntax to force using the primary key index?
</details>
Yes, the primary key columns are duplicated because it makes things much easier. In most cases we could figure out which column indexes are the primaries and use those. However in the case of a table without primary key, simply selecting `*` won't return the primary key. So instead of dealing with the hidden vs not hidden PK case, it's simpler to do it this way. We could optimize this in the future if it becomes a problem.

Yes, we can specify the index. Done.

---

*[cli/dump.go, line 55 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7511#-KLW-hBVZ_u5yxjHI0Eq:-KLXQgc7l6ByhyXP-N-Q:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d4eb7c7183ff1476bae4df7104fe683be2fe3ec/cli/dump.go#L55)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> random local variable
</details>
Done.

---

*[cli/dump.go, line 56 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7511#-KLW-jARdFMn88xb2ZnD:-KLXQkOhFlHrCwAz5Db1:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d4eb7c7183ff1476bae4df7104fe683be2fe3ec/cli/dump.go#L56)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> return dumpTable...
</details>
Done.

---

*[cli/dump.go, line 74 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7511#-KLXVm0CkD-Gs_tUiKYU:-KLXZ9OwZzImE-fSeZEE:-1793423471) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d4eb7c7183ff1476bae4df7104fe683be2fe3ec/cli/dump.go#L74)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> missing verb.
</details>
Poorly phrased. Improved.

---

*[cli/dump.go, line 76 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7511#-KLVzWgF_DdQceBLB20K:-KLXRI1wPx1unIvUOg3q:718764204) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d4eb7c7183ff1476bae4df7104fe683be2fe3ec/cli/dump.go#L76)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> why doesn't this use conn.Begin?
</details>
Because QueryRow and other functions aren't defined on the result of conn.Begin, and BEGIN is already used in other places (zone.go, for example). Also, it's `conn.conn.Begin()`.

---

*[cli/dump.go, line 90 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7511#-KLW02tJ7wbSpNpE7Tvg:-KLZ_oJgCpM3n_am0xAv:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d4eb7c7183ff1476bae4df7104fe683be2fe3ec/cli/dump.go#L90)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> wouldn't it be easier to query system tables and unmarshal protos?
</details>
Done.

---

*[cli/dump.go, line 104 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7511#-KLW-wjoGanEVy8D4sjK:-KLXSDUgMdZbbs1czMVZ:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d4eb7c7183ff1476bae4df7104fe683be2fe3ec/cli/dump.go#L104)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> vals[1]?
</details>
Done.

---

*[cli/dump.go, line 109 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7511#-KLW-xuL0f6qZCHEmX0Z:-KLXSEsoldRlQIqjH9In:-514150589) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d4eb7c7183ff1476bae4df7104fe683be2fe3ec/cli/dump.go#L109)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> vals[4]?
</details>
Done. -_- I'm curious where these bogus indexes came from.

---

*[cli/dump.go, line 206 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7511#-KLXWHO52f9u9izWOes_:-KLXZD4LeSjX0KZKpXmx:1740281843) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d4eb7c7183ff1476bae4df7104fe683be2fe3ec/cli/dump.go#L206)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Good assertion, but this is really not possible unless there's a bug.
</details>
Removed.

---

*[cli/dump.go, line 247 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7511#-KLXXS7E_5mFt4PvJ5jC:-KLa9F3yKBY6wLJYHcZF:1370697872) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d4eb7c7183ff1476bae4df7104fe683be2fe3ec/cli/dump.go#L247)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> The part of the code below makes me wonder why there isn't a standardized facility through which values (given a coltype) represent themselves as a string which can be used in SQL inserts.
</details>
Agree. Not sure we should address that here though.

---

*[cli/dump.go, line 93 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7511#-KLYrs0Y9serMSO1PHoq:-KLZ_pV7Cl9FhBf38dwN:1187967434) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ccd39cab449d9dc326eeb9c56bbe64d8918b681/cli/dump.go#L93)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Whoa!
</details>
We're like a real database now.

---

*[cli/dump_test.go, line 168 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7511#-KLNnLpcbkO9hZGbGENh:-KLXW_fgufQyTZxMVPz9:1992923249) ([raw file](https://github.com/cockroachdb/cockroach/blob/8cd9b7b9426d4c369c7da88f54fc15880c3c96fe/cli/dump_test.go#L168)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Looks like the comment is simply stale. It runs 10 iterations, doesn't it?
</details>
Yes, that comment was stale. But I've done the 1s with a flag thing because it's better.

---

*[cli/dump_test.go, line 189 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7511#-KLXWrp97QtT7omAmQLd:-KLXab7s4Djcsz3C4p5Q:-216615677) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d4eb7c7183ff1476bae4df7104fe683be2fe3ec/cli/dump_test.go#L189)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> `rng, seed := randutil.NewPseudoRand()`
> 
> I don't know here you print it. I'd just print it right here via `log.Infof` (not `t.Logf`, these can get lost if a panic happens iirc).
</details>
Changed.

However I can't figure out how to see the output of log.Infof, so I'm leaving it as t.Logf for now.

---

*[cli/sql_util.go, line 81 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7511#-KLW-M3dE6efHYnW0qIZ:-KLXWerAjsENfZTAzWY_:1532280423) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d4eb7c7183ff1476bae4df7104fe683be2fe3ec/cli/sql_util.go#L81)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> I think this wants to be `args ...driver.Value`
</details>
`[]driver.Value` is the same as `Query`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7511)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
log.Infof should be visible when you run with -v (or the test fails).

On Fri, Jul 1, 2016, 07:13 Matt Jibson <notifications@github.com> wrote:

> Review status: 5 of 7 files reviewed at latest revision, 17 unresolved
> discussions, some commit checks pending.
> ------------------------------
>
> *cli/dump.go, line 34 [r1]
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/7511#-KLNjd2AY3lw7bySVS0o:-KLXQ6G86QBoDlvlYZ1d:-209922296>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/8cd9b7b9426d4c369c7da88f54fc15880c3c96fe/cli/dump.go#L34>):*
> *Previously, petermattis (Peter Mattis) wrote…*
>
> Should this be <database>.<table> to match up with qualified table
> specification in SQL?
>
>
> In the future I think omitting the table name will dump an entire
> database, so leaving it as a separate argument makes it easy to detect that
> case. pg_dump also separates out database and table names, so I think this
> is fine.
>
> ------------------------------
>
> *cli/dump.go, line 55 [r1]
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/7511#-KLNjMpG7bloRRr8tNWY:-KLXQgxk8ZX4HHz6JboL:-499146009>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/8cd9b7b9426d4c369c7da88f54fc15880c3c96fe/cli/dump.go#L55>):*
> *Previously, petermattis (Peter Mattis) wrote…*
>
> The w variable isn't providing much value. Why not just pass os.Stdout as
> the first parameter to dumpTable?
>
>
> Done.
>
> ------------------------------
>
> *cli/dump.go, line 89 [r1]
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/7511#-KLNlmVR7HsXFnS3jdRT:-KLa9D_QeTA-wSxrBgTF:-499146009>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/8cd9b7b9426d4c369c7da88f54fc15880c3c96fe/cli/dump.go#L89>):*
> *Previously, petermattis (Peter Mattis) wrote…*
>
> Given this and the code to fetch the primary key below, I wonder if it
> would be better to read the table descriptor from the system.descriptors
> table. The query would be mildly complicated, but not horrendously so. We
> have code in zone.go to do queries against those tables already, so you'd
> be following existing patterns.
>
>
> Done.
>
> ------------------------------
>
> *cli/dump.go, line 154 [r1]
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/7511#-KLNkAeqDPbC5mnLpUbO:-KLXdAgRcHNeQn50cEBW:1775281126>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/8cd9b7b9426d4c369c7da88f54fc15880c3c96fe/cli/dump.go#L154>):*
> *Previously, petermattis (Peter Mattis) wrote…*
>
> This is selecting the primary key columns and all of the columns, so the
> primary key columns are duplicated, right?
>
> Should we be using the <table>@<index> syntax to force using the primary
> key index?
>
>
> Yes, the primary key columns are duplicated because it makes things much
> easier. In most cases we could figure out which column indexes are the
> primaries and use those. However in the case of a table without primary
> key, simply selecting * won't return the primary key. So instead of
> dealing with the hidden vs not hidden PK case, it's simpler to do it this
> way. We could optimize this in the future if it becomes a problem.
>
> Yes, we can specify the index. Done.
> ------------------------------
>
> *cli/dump.go, line 55 [r2]
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/7511#-KLW-hBVZ_u5yxjHI0Eq:-KLXQgc7l6ByhyXP-N-Q:-499146009>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/1d4eb7c7183ff1476bae4df7104fe683be2fe3ec/cli/dump.go#L55>):*
> *Previously, tamird (Tamir Duberstein) wrote…*
>
> random local variable
>
>
> Done.
>
> ------------------------------
>
> *cli/dump.go, line 56 [r2]
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/7511#-KLW-jARdFMn88xb2ZnD:-KLXQkOhFlHrCwAz5Db1:-499146009>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/1d4eb7c7183ff1476bae4df7104fe683be2fe3ec/cli/dump.go#L56>):*
> *Previously, tamird (Tamir Duberstein) wrote…*
>
> return dumpTable...
>
>
> Done.
>
> ------------------------------
>
> *cli/dump.go, line 74 [r2]
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/7511#-KLXVm0CkD-Gs_tUiKYU:-KLXZ9OwZzImE-fSeZEE:-1793423471>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/1d4eb7c7183ff1476bae4df7104fe683be2fe3ec/cli/dump.go#L74>):*
> *Previously, tschottdorf (Tobias Schottdorf) wrote…*
>
> missing verb.
>
>
> Poorly phrased. Improved.
>
> ------------------------------
>
> *cli/dump.go, line 76 [r2]
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/7511#-KLVzWgF_DdQceBLB20K:-KLXRI1wPx1unIvUOg3q:718764204>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/1d4eb7c7183ff1476bae4df7104fe683be2fe3ec/cli/dump.go#L76>):*
> *Previously, tamird (Tamir Duberstein) wrote…*
>
> why doesn't this use conn.Begin?
>
>
> Because QueryRow and other functions aren't defined on the result of
> conn.Begin, and BEGIN is already used in other places (zone.go, for
> example). Also, it's conn.conn.Begin().
>
> ------------------------------
>
> *cli/dump.go, line 90 [r2]
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/7511#-KLW02tJ7wbSpNpE7Tvg:-KLZ_oJgCpM3n_am0xAv:-499146009>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/1d4eb7c7183ff1476bae4df7104fe683be2fe3ec/cli/dump.go#L90>):*
> *Previously, tamird (Tamir Duberstein) wrote…*
>
> wouldn't it be easier to query system tables and unmarshal protos?
>
>
> Done.
>
> ------------------------------
>
> *cli/dump.go, line 104 [r2]
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/7511#-KLW-wjoGanEVy8D4sjK:-KLXSDUgMdZbbs1czMVZ:-499146009>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/1d4eb7c7183ff1476bae4df7104fe683be2fe3ec/cli/dump.go#L104>):*
> *Previously, tamird (Tamir Duberstein) wrote…*
>
> vals[1]?
>
>
> Done.
>
> ------------------------------
>
> *cli/dump.go, line 109 [r2]
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/7511#-KLW-xuL0f6qZCHEmX0Z:-KLXSEsoldRlQIqjH9In:-514150589>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/1d4eb7c7183ff1476bae4df7104fe683be2fe3ec/cli/dump.go#L109>):*
> *Previously, tamird (Tamir Duberstein) wrote…*
>
> vals[4]?
>
>
> Done. -_- I'm curious where these bogus indexes came from.
>
> ------------------------------
>
> *cli/dump.go, line 206 [r2]
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/7511#-KLXWHO52f9u9izWOes_:-KLXZD4LeSjX0KZKpXmx:1740281843>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/1d4eb7c7183ff1476bae4df7104fe683be2fe3ec/cli/dump.go#L206>):*
> *Previously, tschottdorf (Tobias Schottdorf) wrote…*
>
> Good assertion, but this is really not possible unless there's a bug.
>
>
> Removed.
>
> ------------------------------
>
> *cli/dump.go, line 247 [r2]
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/7511#-KLXXS7E_5mFt4PvJ5jC:-KLa9F3yKBY6wLJYHcZF:1370697872>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/1d4eb7c7183ff1476bae4df7104fe683be2fe3ec/cli/dump.go#L247>):*
> *Previously, tschottdorf (Tobias Schottdorf) wrote…*
>
> The part of the code below makes me wonder why there isn't a standardized
> facility through which values (given a coltype) represent themselves as a
> string which can be used in SQL inserts.
>
>
> Agree. Not sure we should address that here though.
>
> ------------------------------
>
> *cli/dump.go, line 93 [r3]
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/7511#-KLYrs0Y9serMSO1PHoq:-KLZ_pV7Cl9FhBf38dwN:1187967434>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/0ccd39cab449d9dc326eeb9c56bbe64d8918b681/cli/dump.go#L93>):*
> *Previously, petermattis (Peter Mattis) wrote…*
>
> Whoa!
>
>
> We're like a real database now.
>
> ------------------------------
>
> *cli/dump_test.go, line 168 [r1]
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/7511#-KLNnLpcbkO9hZGbGENh:-KLXW_fgufQyTZxMVPz9:1992923249>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/8cd9b7b9426d4c369c7da88f54fc15880c3c96fe/cli/dump_test.go#L168>):*
> *Previously, tschottdorf (Tobias Schottdorf) wrote…*
>
> Looks like the comment is simply stale. It runs 10 iterations, doesn't it?
>
>
> Yes, that comment was stale. But I've done the 1s with a flag thing
> because it's better.
>
> ------------------------------
>
> *cli/dump_test.go, line 189 [r2]
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/7511#-KLXWrp97QtT7omAmQLd:-KLXab7s4Djcsz3C4p5Q:-216615677>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/1d4eb7c7183ff1476bae4df7104fe683be2fe3ec/cli/dump_test.go#L189>):*
> *Previously, tschottdorf (Tobias Schottdorf) wrote…*
>
> rng, seed := randutil.NewPseudoRand()
>
> I don't know here you print it. I'd just print it right here via log.Infof
> (not t.Logf, these can get lost if a panic happens iirc).
>
>
> Changed.
>
> However I can't figure out how to see the output of log.Infof, so I'm
> leaving it as t.Logf for now.
> ------------------------------
>
> *cli/sql_util.go, line 81 [r2]
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/7511#-KLW-M3dE6efHYnW0qIZ:-KLXWerAjsENfZTAzWY_:1532280423>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/1d4eb7c7183ff1476bae4df7104fe683be2fe3ec/cli/sql_util.go#L81>):*
> *Previously, tamird (Tamir Duberstein) wrote…*
>
> I think this wants to be args ...driver.Value
>
>
> []driver.Value is the same as Query
>
> ------------------------------
>
> *Comments from Reviewable
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/7511>*
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/7511#issuecomment-229921568>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AE135NhNyIrOyEVCGHoZaldgSzKdJtC0ks5qRPZKgaJpZM4JAeoI>
> .
>
-- 

-- Tobias

tamird commented:
Reviewed 1 of 2 files at r4.
Review status: 6 of 7 files reviewed at latest revision, 12 unresolved discussions, some commit checks failed.

---

*[cli/dump.go, line 117 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7511#-KLaO6x-vHl7AXLN7fdH:-KLaO6x-vHl7AXLN7fdI:518259481) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae85d90f62903bb07acad86dc751244a2fc42b69/cli/dump.go#L117)):*
> ```Go
> 
> 	// Build the SELECT query.
> 	sbuf := bytes.NewBufferString(fmt.Sprintf("SELECT %s, * FROM %s@%s AS OF SYSTEM TIME '%s'", indexes, tablename, primaryIndex, clusterTS))
> ```

style nit: i'd have expected this to read:
```
var sbuf bytes.Buffer
fmt.Fprintf(&sbuf, ...)
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7511)*
<!-- Sent from Reviewable.io -->

mjibson commented:
Review status: 4 of 7 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[cli/dump.go, line 117 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7511#-KLaO6x-vHl7AXLN7fdH:-KLbpes_Ybgk5h277wqx:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/ae85d90f62903bb07acad86dc751244a2fc42b69/cli/dump.go#L117)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> style nit: i'd have expected this to read:
> ```
> var sbuf bytes.Buffer
> fmt.Fprintf(&sbuf, ...)
> ```
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7511)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
LGTM

---

Review status: 4 of 7 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[cli/dump.go, line 78 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7511#-KLrsxgqL0QmNnuIwSdI:-KLrsxgqL0QmNnuIwSdJ:-2051167923) ([raw file](https://github.com/cockroachdb/cockroach/blob/2836c4c4ed5cf60acbce3692ce38ae273f852ddd/cli/dump.go#L78)):*
> ```Go
> 	var clusterTSStart int64
> 
> 	vals, err := conn.QueryRow("SELECT cluster_logical_timestamp()::int", nil)
> ```

By discarding the logical ticks, you'll execute the time travel queries at a (very slightly) different time than the metadata query. Probably worth a TODO or an issue to allow full HLC timestamps as input for time travel. Or you could pick a timestamp on the client and do the metadata reads as of that time too. 

---

*[cli/dump.go, line 186 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7511#-KLrwK5BrJ3h15LIJRNq:-KLrwK5BrJ3h15LIJRNr:977336439) ([raw file](https://github.com/cockroachdb/cockroach/blob/2836c4c4ed5cf60acbce3692ce38ae273f852ddd/cli/dump.go#L186)):*
> ```Go
> 			}
> 			pk = vals[:len(index)]
> 			// Primary key strings need to be converted to a go string, but not SQL
> ```

I don't see a test for string or byte primary keys (unless I'm missing something, all the tests use integer PKs). 

---

*[cli/dump_test.go, line 204 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7511#-KLrwz9mgORMKwThLQog:-KLrwz9mgORMKwThLQoh:1961226845) ([raw file](https://github.com/cockroachdb/cockroach/blob/2836c4c4ed5cf60acbce3692ce38ae273f852ddd/cli/dump_test.go#L204)):*
> ```Go
> 			// Generate a random number of random inserts.
> 			i := rnd.Int63()
> 			f := rnd.Float64()
> ```

This will never generate special values like `NaN` or `Inf` (or `NULL`, which could apply to any column type)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7511#-:-KLrxtL2gt5WEa59DIld:676834233)*
<!-- Sent from Reviewable.io -->

mjibson commented:
Review status: 4 of 7 files reviewed at latest revision, 8 unresolved discussions, some commit checks pending.

---

*[cli/dump.go, line 78 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7511#-KLrsxgqL0QmNnuIwSdI:-KLwOgBoNpeHMyJZrpA0:-534236213) ([raw file](https://github.com/cockroachdb/cockroach/blob/c00e88d7b722e36d4991d24fc94ab0bec6de3f54/cli/dump.go#L78)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> By discarding the logical ticks, you'll execute the time travel queries at a (very slightly) different time than the metadata query. Probably worth a TODO or an issue to allow full HLC timestamps as input for time travel. Or you could pick a timestamp on the client and do the metadata reads as of that time too. 
</details>
Opened #7624 for this.

---

*[cli/dump.go, line 186 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7511#-KLrwK5BrJ3h15LIJRNq:-KLwPPOUwj3y7o0bCAM-:657406640) ([raw file](https://github.com/cockroachdb/cockroach/blob/c00e88d7b722e36d4991d24fc94ab0bec6de3f54/cli/dump.go#L186)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> I don't see a test for string or byte primary keys (unless I'm missing something, all the tests use integer PKs). 
</details>
I made the bytes column the PK on the TestDumpBytes test.

---

*[cli/dump_test.go, line 204 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7511#-KLrwz9mgORMKwThLQog:-KLwQtU0zPjPt6fjst0v:-136407847) ([raw file](https://github.com/cockroachdb/cockroach/blob/c00e88d7b722e36d4991d24fc94ab0bec6de3f54/cli/dump_test.go#L204)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> This will never generate special values like `NaN` or `Inf` (or `NULL`, which could apply to any column type)
</details>
I've added NULL, Inf, and NaN tests elsewhere.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7511)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 2 of 3 files at r5, 2 of 2 files at r7.
Review status: all files reviewed at latest revision, 16 unresolved discussions, some commit checks failed.

---

*[cli/dump.go, line 74 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7511#-KLXVm0CkD-Gs_tUiKYU:-KLxsmqvMROjsSScxmww:1618789527) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d4eb7c7183ff1476bae4df7104fe683be2fe3ec/cli/dump.go#L74)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary>
> Poorly phrased. Improved.
</details>
missing article

---

*[cli/dump.go, line 97 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7511#-KLxsw7_lZIT0yfNs593:-KLxsw7_lZIT0yfNs594:-857012764) ([raw file](https://github.com/cockroachdb/cockroach/blob/ecc5930db97637a3b2c7c477b93b7dc486f5cd62/cli/dump.go#L97)):*
> ```Go
> 		[]driver.Value{origTableName, origDBName})
> 	if err == io.EOF {
> 		return fmt.Errorf("unknown table")
> ```

can't it be an unknown database as well? why not log both names?

---

*[cli/dump.go, line 104 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7511#-KLxt3vyY0eraudGRWJx:-KLxt3vzIln7kVZcS8rA:166874) ([raw file](https://github.com/cockroachdb/cockroach/blob/ecc5930db97637a3b2c7c477b93b7dc486f5cd62/cli/dump.go#L104)):*
> ```Go
> 	b := row[0].([]byte)
> 	desc := &sqlbase.Descriptor{}
> 	println("LB", len(b))
> ```

?

---

*[cli/dump.go, line 106 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7511#-KLxtAjOgylA-qnzzeLJ:-KLxtAjOgylA-qnzzeLK:1258202887) ([raw file](https://github.com/cockroachdb/cockroach/blob/ecc5930db97637a3b2c7c477b93b7dc486f5cd62/cli/dump.go#L106)):*
> ```Go
> 	println("LB", len(b))
> 	if err := proto.Unmarshal(b, desc); err != nil {
> 		return err
> ```

throughout: these errors are about to be user-facing, let's `errors.Wrapf` them

---

*[cli/dump.go, line 108 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7511#-KLxt4p6WjjSz3p8ZjN-:-KLxt4p6WjjSz3p8ZjN0:166874) ([raw file](https://github.com/cockroachdb/cockroach/blob/ecc5930db97637a3b2c7c477b93b7dc486f5cd62/cli/dump.go#L108)):*
> ```Go
> 		return err
> 	}
> 	println("UMD")
> ```

?

---

*[cli/dump.go, line 111 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7511#-KLxt7stkJy4lenIuGPJ:-KLxt7stkJy4lenIuGPK:229863670) ([raw file](https://github.com/cockroachdb/cockroach/blob/ecc5930db97637a3b2c7c477b93b7dc486f5cd62/cli/dump.go#L111)):*
> ```Go
> 	table := desc.GetTable()
> 	if table == nil {
> 		return fmt.Errorf("expected table descriptor")
> ```

not a useful error message

---

*[cli/dump.go, line 241 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7511#-KLxtWktk-YMgSnS7mXM:-KLxtWktk-YMgSnS7mXN:166874) ([raw file](https://github.com/cockroachdb/cockroach/blob/ecc5930db97637a3b2c7c477b93b7dc486f5cd62/cli/dump.go#L241)):*
> ```Go
> 			i++
> 		}
> 		fmt.Fprintln(os.Stderr, "NPK", nextPK)
> ```

?

---

*[cli/dump.go, line 263 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7511#-KLxtYFbPXFSBAm8k52L:-KLxtYFbPXFSBAm8k52M:166874) ([raw file](https://github.com/cockroachdb/cockroach/blob/ecc5930db97637a3b2c7c477b93b7dc486f5cd62/cli/dump.go#L263)):*
> ```Go
> 			return err
> 		}
> 		fmt.Fprintln(os.Stderr, "PK", pk, "\n")
> ```

?

---

*[cli/dump_test.go, line 290 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7511#-KLxtbxuQKqeB3ikiYmn:-KLxtbxvpRj4wTtLYmx6:166874) ([raw file](https://github.com/cockroachdb/cockroach/blob/ecc5930db97637a3b2c7c477b93b7dc486f5cd62/cli/dump_test.go#L290)):*
> ```Go
> 		}
> 		dump := buf.String()
> 		fmt.Println(dump)
> ```

?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7511)*
<!-- Sent from Reviewable.io -->

mjibson commented:
Changes in most recent revision that may warrant review: rows.Next (for all sub commands) now performs a copy of any `[]byte` value, because those can get mutated during calls to rows.Next or rows.Close. QueryRow has been simplified to wrap Query.

---

Review status: 4 of 7 files reviewed at latest revision, 16 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7511#-:-KLyu8JE4PaljZqhU397:535134021)*
<!-- Sent from Reviewable.io -->

mjibson commented:
Review status: 4 of 7 files reviewed at latest revision, 16 unresolved discussions, some commit checks pending.

---

*[cli/dump.go, line 74 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7511#-KLXVm0CkD-Gs_tUiKYU:-KLywdqaDlhBuCkvEa79:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d4eb7c7183ff1476bae4df7104fe683be2fe3ec/cli/dump.go#L74)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> missing article
</details>
Done.

---

*[cli/dump.go, line 97 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7511#-KLxsw7_lZIT0yfNs593:-KLywpxDJgapLzXKIT0o:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/ecc5930db97637a3b2c7c477b93b7dc486f5cd62/cli/dump.go#L97)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> can't it be an unknown database as well? why not log both names?
</details>
Done.

---

*[cli/dump.go, line 104 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7511#-KLxt3vyY0eraudGRWJx:-KLywr5yirqVz3BIR5a4:1731980910) ([raw file](https://github.com/cockroachdb/cockroach/blob/ecc5930db97637a3b2c7c477b93b7dc486f5cd62/cli/dump.go#L104)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> ?
</details>
Debug. Removed.

---

*[cli/dump.go, line 106 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7511#-KLxtAjOgylA-qnzzeLJ:-KLyxFoJcNomKIE6dcWV:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/ecc5930db97637a3b2c7c477b93b7dc486f5cd62/cli/dump.go#L106)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> throughout: these errors are about to be user-facing, let's `errors.Wrapf` them
</details>
Done.

---

*[cli/dump.go, line 108 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7511#-KLxt4p6WjjSz3p8ZjN-:-KLywuSxNDkHgWr2FgsI:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/ecc5930db97637a3b2c7c477b93b7dc486f5cd62/cli/dump.go#L108)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> ?
</details>
Done.

---

*[cli/dump.go, line 111 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7511#-KLxt7stkJy4lenIuGPJ:-KLyxUSf6KqSZryFmDoa:-2076588413) ([raw file](https://github.com/cockroachdb/cockroach/blob/ecc5930db97637a3b2c7c477b93b7dc486f5cd62/cli/dump.go#L111)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> not a useful error message
</details>
I'm really not sure what to put here instead. I've added "internal error", but if this happens it's not a user error, it's something on our end.

---

*[cli/dump.go, line 241 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7511#-KLxtWktk-YMgSnS7mXM:-KLyxd00HiHHu7AnRczM:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/ecc5930db97637a3b2c7c477b93b7dc486f5cd62/cli/dump.go#L241)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> ?
</details>
Done.

---

*[cli/dump.go, line 263 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7511#-KLxtYFbPXFSBAm8k52L:-KLyxdnXqQXav-KSzQMl:2030846982) ([raw file](https://github.com/cockroachdb/cockroach/blob/ecc5930db97637a3b2c7c477b93b7dc486f5cd62/cli/dump.go#L263)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> ?
</details>
This diff you reviewed was my mid-work commit so I could resume working at home. Was not meant for review.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7511)*
<!-- Sent from Reviewable.io -->

mjibson commented:
Review status: 4 of 7 files reviewed at latest revision, 16 unresolved discussions, some commit checks pending.

---

*[cli/dump_test.go, line 290 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7511#-KLxtbxuQKqeB3ikiYmn:-KLyyRRA3IXmc2xAKtgv:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/ecc5930db97637a3b2c7c477b93b7dc486f5cd62/cli/dump_test.go#L290)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> ?
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7511)*
<!-- Sent from Reviewable.io -->

madelynnblue committed:
sql: add dump subcommand

Introspect over a specified table to determine its structure. Iteratively
select all rows from the table in pages using time travel queries to
guarantee they don't change.

TestDumpRandom was run for 20 minutes and didn't report any unexpected
values, so I have fairly high confidence that the bytes a user has stored
will correctly be inserted if the dump file is piped into the sql subcommand.

jseldess commented:
Docs: https://www.cockroachlabs.com/docs/back-up-and-restore-data.html
